numpy/core/src/multiarray/common_dtype.c

```diff
@@ -0,0 +1,318 @@
+                 */
+                PyArray_DTypeMeta *tmp = dtypes[low];
+                dtypes[low] = dtypes[high];
+                dtypes[high] = tmp;
+            }
+
+            Py_XSETREF(res, dtypes[low]->common_dtype(dtypes[low], dtypes[high]));
+            if (res == NULL) {
+                return NULL;
+            }
+        }
+
+        if (res == (PyArray_DTypeMeta *)Py_NotImplemented) {
+            PyArray_DTypeMeta *tmp = dtypes[low];
+            dtypes[low] = dtypes[high];
+            dtypes[high] = tmp;
+        }
+        if (res == dtypes[low]) {
+            /* `dtypes[high]` cannot influence the final result, so clear: */
+            dtypes[high] = NULL;
+        }
+    }
+
+    if (length == 2) {
+        return res;
+    }
+    Py_DECREF(res);
+    return reduce_dtypes_to_most_knowledgeable(length - half, dtypes);
+}
+
+
+/**
+ * Promotes a list of DTypes with each other in a way that should guarantee
+ * stable results even when changing the order.
+ *
+ * In general this approach always works as long as the most generic dtype
+ * is either strictly larger, or compatible with all other dtypes.
+ * For example promoting float16 with any other float, integer, or unsigned
+ * integer again gives a floating point number. And any floating point number
+ * promotes in the "same way" as `float16`.
+ * If a user inserts more than one type into the NumPy type hierarchy, this
+ * can break. Given:
+ *     uint24 + int32 -> int48  # Promotes to a *new* dtype!
+ *
+ * The following becomes problematic (order does not matter):
+ *         uint24 +      int16  +           uint32  -> int64
+ *    <==      (uint24 + int16) + (uint24 + uint32) -> int64
+ *    <==                int32  +           uint32  -> int64
+ *
+ * It is impossible to achieve an `int48` result in the above.
+ *
+ * This is probably only resolvable by asking `uint24` to take over the
+ * whole reduction step; which we currently do not do.
+ * (It may be possible to notice the last up-cast and implement use something
+ * like: `uint24.nextafter(int32).__common_dtype__(uint32)`, but that seems
+ * even harder to grasp.)
+ *
+ * Note that a case where two dtypes are mixed (and know nothing about each
+ * other) will always generate an error:
+ *     uint24 + int48 + int64 -> Error
+ *
+ * Even though `int64` is a safe solution, since `uint24 + int64 -> int64` and
+ * `int48 + int64 -> int64` and `int64` and there cannot be a smaller solution.
+ *
+ * //TODO: Maybe this function should allow not setting an error?
+ *
+ * @param length Number of dtypes (and values) must be at least 1
+ * @param dtypes The concrete or abstract DTypes to promote
+ * @return NULL or the promoted DType.
+ */
+NPY_NO_EXPORT PyArray_DTypeMeta *
+PyArray_PromoteDTypeSequence(
+        npy_intp length, PyArray_DTypeMeta **dtypes_in)
+{
+    if (length == 1) {
+        Py_INCREF(dtypes_in[0]);
+        return dtypes_in[0];
+    }
```

Comment:
test missing?

Reply:
I think this one is impossible to reach from Python currently.  A bit of a bummer, but I would like to be pragmatic about it.

Finally, got the rest of the coverage right though.  (I was missing multiple Python scalars as inputs paths, probably that also were not also _all_ scalars).